packages/components/src/components/pin-code/pin-code.tsx

```diff
@@ -0,0 +1,218 @@
+import {
+  AllowedTypes,
+  attachComponentCss,
+  FORM_STATES,
+  hasDescription,
+  hasLabel,
+  hasMessage,
+  hasPropValueChanged,
+  isWithinForm,
+  THEMES,
+  validateProps,
+} from '../../utils';
+import { getComponentCss } from './pin-code-styles';
+import {
+  initHiddenInput,
+  inputIsSingleDigit,
+  joinInputValues,
+  PIN_CODE_TYPES,
+  syncHiddenInput,
+} from './pin-code-utils';
+import { StateMessage } from '../common/state-message/state-message';
+import { Required } from '../common/required/required';
+
+const propTypes: PropTypes<typeof PinCode> = {
+  label: AllowedTypes.string,
+  description: AllowedTypes.string,
+  length: AllowedTypes.number,
+  hideLabel: AllowedTypes.breakpoint('boolean'),
+  state: AllowedTypes.oneOf<PinCodeState>(FORM_STATES),
+  disabled: AllowedTypes.boolean,
+  required: AllowedTypes.boolean,
+  message: AllowedTypes.string,
+  theme: AllowedTypes.oneOf<Theme>(THEMES),
+  type: AllowedTypes.oneOf<PinCodeType>(PIN_CODE_TYPES),
+  value: AllowedTypes.oneOf<ValidatorFunction>([AllowedTypes.string, AllowedTypes.number]),
+};
+
+@Component({
+  tag: 'p-pin-code',
+  shadow: { delegatesFocus: true },
+})
+export class PinCode {
+  @Element() public host!: HTMLElement;
+
+  /** The description text. */
+  @Prop() public description?: string = '';
+
+  /** The label text. */
+  @Prop() public label?: string = '';
+
+  /** Number of characters of the Pin Code. */
+  @Prop() public length?: number = 4;
+
+  /** Show or hide label and description text. For better accessibility it is recommended to show the label. */
+  @Prop() public hideLabel?: BreakpointCustomizable<boolean> = false;
+
+  /** The validation state. */
+  @Prop() public state?: PinCodeState = 'none';
+
+  /** Disables the Pin Code. No events will be triggered while disabled state is active. */
+  @Prop() public disabled?: boolean = false;
+
+  /** Marks the Pin Code as required. */
+  @Prop() public required?: boolean = false;
+
+  /** The message styled depending on validation state. */
+  @Prop() public message?: string = '';
+
+  /** Pin Code type. */
+  @Prop() public type?: PinCodeType = 'number';
+
+  /** Sets the initial value of the Pin Code. */
+  @Prop() public value?: string = '';
+
+  /** Emitted when selected element changes. */
+  @Event({ bubbles: false }) public update: EventEmitter<PinCodeUpdateEvent>;
+
+  /** Adapts the color depending on the theme. */
+  @Prop() public theme?: Theme = 'light';
+
+  private isWithinForm: boolean;
+  private hiddenInput: HTMLInputElement;
+  private pinCodeElements: HTMLInputElement[] = [];
+  // TODO: private ariaElement: HTMLSpanElement;
+
+  public connectedCallback(): void {
+    this.isWithinForm = isWithinForm(this.host);
+  }
+
+  public componentWillLoad(): void {
+    // make sure initial value is not longer than pin code length
+    if (this.value) {
+      this.value.toString().slice(0, this.length);
```

Comment:
I think `this.value` is still the same because the result of this manipulation is never saved.